### PR TITLE
Target-specific builds for Windows

### LINKED ISSUE #15: Unable to build example
I installed latest github version of the package and tried creating the following rust function from the README:
```R
rust_function("fn add(a:f64, b:f64) -> f64 { a + b }")
```
I unfortuantely experienced a liking issue at the end, see attached log:
<details>

```pwsh
build directory: C:\Users\[redacted]\AppData\Local\Temp\Rtmp04R3wM\file46582e2c1e4d

    Updating git repository `https://github.com/extendr/extendr`
    Updating crates.io index
   Compiling winapi-build v0.1.1
   Compiling winapi v0.3.9
   Compiling winapi v0.2.8
   Compiling proc-macro2 v1.0.24
   Compiling unicode-xid v0.2.1
   Compiling syn v1.0.58
   Compiling extendr-engine v0.1.11 (https://github.com/extendr/extendr#f714309f)
   Compiling lazy_static v1.4.0
   Compiling kernel32-sys v0.2.2
   Compiling quote v1.0.8
   Compiling extendr-macros v0.1.11 (https://github.com/extendr/extendr#f714309f)
   Compiling libR-sys v0.2.1
   Compiling extendr-api v0.1.11 (https://github.com/extendr/extendr#f714309f)
   Compiling rextendr1 v0.0.1 (C:\Users\[redacted]\AppData\Local\Temp\Rtmp04R3wM\file46582e2c1e4d)
error: linking with `link.exe` failed: exit code: 1181
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "/NXCOMPAT" "/LIBPATH:C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.1109b7oki3iy8t63.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.147sjo61quy01gmx.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.18c95s8sfqsxquuj.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.1ibpcj3kfkp6aui2.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.1lqos2f8068mw4qh.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.26e1qwh59i1q5m7k.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.28pcu9knbknhvzkw.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.2c9dqcehnk29gg5h.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.2j9zzh0cm8tm34dr.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.2lh12q8e2e0bj40g.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.2mquew07qtassp2j.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.2zt21rjm9icixcp2.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.328vlxktswt53wtg.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.33qkn1pkb117k5tg.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.3e9fnqq3jbqxzfk9.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.3hkoh4ve50ze28sj.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.3q25qttx9oje29z4.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.3rg0lshqq2jwm9lu.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.446tnh1hgeolf4fm.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.46z1eys4ytheu3u7.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.4fdhf75i23bk8vxa.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.4x2o391x3w6z5ml2.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.4zaytgzpph72bfw6.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.59upnde7k2408fk1.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.60phqctaujov6k4.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.br35nq041eolkor.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.f9qxssc7l1rmtgz.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.meaiqnde68um5p1.rcgu.o" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.nykpq4yd0pxi0xe.rcgu.o" "/OUT:C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.dll" "/DEF:C:\\Users\\[redacted]\\AppData\\Local\\Temp\\rustcu1SLss\\lib.def" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.45yv15xwvrrq7lsi.rcgu.o" "/OPT:REF,NOICF" "/DLL" "/IMPLIB:C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps\\rextendr1.dll.lib" "/DEBUG" "/NATVIS:C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libstd.natvis" "/LIBPATH:C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d/target\\debug\\deps" "/LIBPATH:C:/PROGRA~1/R/R-40~1.3\\bin\\x64" "/LIBPATH:C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d\\target\\debug\\deps\\libextendr_api-e53357031a4dd514.rlib" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d\\target\\debug\\deps\\libextendr_engine-5644b2bd52a06f57.rlib" "C:\\Users\\[redacted]\\AppData\\Local\\Temp\\Rtmp04R3wM\\file46582e2c1e4d\\target\\debug\\deps\\liblibR_sys-f9732ac30d4d11f1.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd-f771f8d0374ceeb1.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libpanic_unwind-9049f341d2774a65.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_demangle-6a47dea777db971c.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libhashbrown-a3b06009982e3d9c.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_alloc-ebb710aea302de49.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libunwind-4304b0c1525af3c3.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcfg_if-c70f620fc91130c5.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liblibc-3b06f844280b3802.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liballoc-86edfc7ab798fc86.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_core-3aa1c50c964075c2.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcore-3ec787c945bbba26.rlib" "C:\\Users\\[redacted]\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-937aede5bfe8081e.rlib" "R.lib" "advapi32.lib" "ws2_32.lib" "userenv.lib" "msvcrt.lib"
  = note: LINK : fatal error LNK1181: cannot open input file 'R.lib'
          

error: aborting due to previous error

error: could not compile `rextendr1`

To learn more, run the command again with --verbose.
Error: Rust code could not be compiled successfully. Aborting.
Execution halted

```

</details>

This happened at least on Windows. One of the problems I see is that the linking is attempted agains `R.lib`, while it should be `R.dll`. It seems that either `Makevars` or something else is poorly configured.

With that said, maybe we should have 'real' tests attached to this repository? For instance, creating such `add` function and checking its output, because the package can be installed and `testthat` succeeds, yet I am unable to use it.

### PR DESCRIPTION
**Pull request by Ilia-Kosenkov**
Fixes #15 for Windows.

### CODE REVIEW

**Comment by Ilia-Kosenkov** [comment]
We have some Notes generated by R CMD CHECK:
https://github.com/extendr/rextendr/pull/16/checks?check_run_id=1681742389#step:9:121

**Comment by clauswilke** [comment]
Is the change to the `patch.crates_io` necessary to make things work on Windows? I'm not sure I like it, it makes writing the specification more complicated, and as a consequence, more error-prone.

**Comment by Ilia-Kosenkov** [comment]
@clauswilke, No, but it actually affected my editor's parsing capabilities because of nested single/double quotes and curly brackets. This can be obviously reversed or discussed elsewhere. I would prefer to have a key-value collection of crates rather than fully stringified parameters.

**Comment by clauswilke** [comment]
> I would prefer to have a key-value collection of crates rather than fully stringified parameters.

The R way of doing this is probably to pass on unnamed strings as is and interpret named strings as key-value pairs. That allows for both uses. Feel free to create a separate PR implementing this.

**Comment by Ilia-Kosenkov** [comment]
@clauswilke ,
Anyway I removed this change.
Please double check the rest.
These modifications solve issues on my Windows and do not create any regressions in WSL2.

File: R/source.R [hunk 1]
```diff
@@ -149,7 +153,16 @@ rust_source <- function(file, code = NULL, dependencies = NULL,
     paste0("lib", libname, get_dynlib_ext())
   }
 
-  shared_lib <- file.path(dir, "target", ifelse(profile == "dev", "debug", "release"), libfilename)
+  target_folder <- ifelse(
+    is.null(specific_target),
+    "target",
+    sprintf("target%s%s", .Platform$file.sep, specific_target))
```
Comment (clauswilke) [inline R/source.R:159]: Please move the last closing parenthesis onto a new line.

File: R/source.R [hunk 2]
```diff
@@ -197,6 +210,21 @@ get_dynlib_ext <- function() {
...
 
+get_specific_target_name <- function() {
+  sysinf <- Sys.info()
+
+  if  (!is.null(sysinf) && sysinf["sysname"] == "Windows") {
+    if (R.version$arch == "x86_64")
+      return("x86_64-pc-windows-gnu")
+
+    if (R.version$arch == "i386")
+      return("i686-pc-windows-gnu")
```
Comment (clauswilke) [inline R/source.R:221]: Here and above: Please enclose the `return()` statement into curly braces, as described here: https://style.tidyverse.org/syntax.html#inline-statements

**Review by clauswilke — COMMENTED** [review]
Generally looks good to me. I see a few minor formatting issues.

**Comment by clauswilke** [comment]
Thanks!

**Comment by clauswilke** [comment]
All the Notes should be fixed now.

**Pull request merged**